### PR TITLE
Discover: Add deployed method field to deploy service event

### PR DESCRIPTION
**Pull request by kimlisa**
part of https://github.com/gravitational/teleport/issues/22132
cloud PR (merge first, `teleport.proto` has to be copy pasta): https://github.com/gravitational/cloud/pull/5057

We now have two methods of agent deployment in the web UI:
  - auto: teleport automagically deploys an agent for user
  - manually: user has their own instance where they run the download script
  - unspecified: when deploying agent is skipped due to an agent already existing

### CODE REVIEW

**Review by marcoandredinis — APPROVED** [review]
Just left a nit on the cloud's PR

File: api/proto/teleport/usageevents/v1/usageevents.proto [hunk 1]
```diff
@@ -227,6 +227,20 @@ message UIDiscoverDeployServiceEvent {
   DiscoverMetadata metadata = 1;
   DiscoverResourceMetadata resource = 2;
   DiscoverStepStatus status = 3;
+
+  // DeployedMethod desribes the method used to deploy a service.
```
Comment (ravicious) [inline api/proto/teleport/usageevents/v1/usageevents.proto:231]:
```suggestion
  // DeployedMethod describes the method used to deploy a service.
```

File: api/proto/teleport/usageevents/v1/usageevents.proto [hunk 2]
```diff
@@ -227,6 +227,20 @@ message UIDiscoverDeployServiceEvent {
...
+  enum DeployedMethod {
+    // DEPLOY_METHOD_UNSPECIFIED means there was an existing service
+    // so deploying step got skipped.
+    DEPLOYED_METHOD_UNSPECIFIED = 0;
```
Comment (ravicious) [inline api/proto/teleport/usageevents/v1/usageevents.proto:235]: Deployed method or deploy method? FWIW, deploy method sounds better to me.

**Review by ravicious — APPROVED** [review]
Leaving a preemptive approval, though I think the naming needs to be made consistent.

In the cloud PR, there's these two lines:

```go
discoverDeployedServiceMethodProperty posthog.EventProperty = "tp.discover.deployed_service.method"
// …
e.AddProperty(discoverDeployedServiceMethodProperty, oneof.UiDiscoverDeployServiceEvent.DeployedMethod.String())
```

I feel like "deployed_service.method", but "deployed_method" should be "deploy_method". But again, I don't have context on this, so YMMV.

File: web/packages/teleport/src/services/userEvent/types.ts [hunk 3]
```diff
@@ -175,13 +175,23 @@ export type DiscoverEventData = DiscoverEventStepStatus & {
...
+  // event and describes how a agent got deployed.
+  serviceDeployedMethod?: DiscoverServiceDeployMethod;
 };
 
 export type DiscoverEventStepStatus = {
   stepStatus: DiscoverEventStatus;
   stepStatusError?: string;
 };
 
+export enum DiscoverServiceDeployMethod {
```
[the remark this answers is not in the document]
Comment (kimlisa) [reply web/packages/teleport/src/services/userEvent/types.ts:194]: sorry about that, i originally started with `deployXXX`, then found myself naming things `deployedXXX` and decided to go with that instead and thought I renamed everything. I went back to `deployXXX`

File: web/packages/teleport/src/services/userEvent/types.ts [hunk 4]
```diff
@@ -175,13 +175,34 @@ export type DiscoverEventData = DiscoverEventStepStatus & {
...
+  Unspecified = 'DEPLOY_METHOD_UNSPECIFIED',
+  Auto = 'DEPLOY_METHOD_AUTO',
+  Manual = 'DEPLOY_METHOD_MANUAL',
+}
+
+export enum DiscoverServiceDeployType {
+  Unspecified = 'DEPLOY_TYPE_UNSPECIFIED',
+  InstallScript = 'DEPLOY_TYPE_INSTALL_SCRIPT',
+  AmazonEcs = 'DEPLOY_TYPE_AMAZON_ECS',
```
Comment (marcoandredinis) [inline web/packages/teleport/src/services/userEvent/types.ts:203]: This should be `AmazonECS` unless we need it to be `Ecs` for some reason

**Review by marcoandredinis — APPROVED** [review]
Just left a nit

File: lib/usagereporter/web/userevent_discover_test.go [hunk 5]
```diff
@@ -114,6 +114,31 @@ func TestDiscoverEventDataToUsageEvent(t *testing.T) {
 				},
 			}},
 		},
+		{
+			name:     uiDiscoverDeployServiceEvent + "/success_test_deployed_method_setting",
```
Comment (ravicious) [inline lib/usagereporter/web/userevent_discover_test.go:118]: That's one of the two "deployed method" that `rg "deployed.?method"` has returned. ;)

```suggestion
			name:     uiDiscoverDeployServiceEvent + "/success_test_deploy_method_setting",
```

File: lib/usagereporter/web/userevent_discover.go [hunk 6]
```diff
@@ -116,11 +126,21 @@ func (d *DiscoverEventData) ToUsageEvent(eventName string) (*usageeventsv1.Usage
 		}}, nil
 
 	case uiDiscoverDeployServiceEvent:
+		deployMethodEnum, ok := usageeventsv1.UIDiscoverDeployServiceEvent_DeployMethod_value[d.ServiceDeploy.Method]
+		if !ok {
+			return nil, trace.BadParameter("invalid deployed method %s", d.ServiceDeploy.Method)
```
Comment (ravicious) [inline lib/usagereporter/web/userevent_discover.go:131]:
```suggestion
			return nil, trace.BadParameter("invalid deploy method %s", d.ServiceDeploy.Method)
```

File: lib/usagereporter/web/userevent_discover.go [hunk 7]
```diff
@@ -116,11 +126,21 @@ func (d *DiscoverEventData) ToUsageEvent(eventName string) (*usageeventsv1.Usage
...
+		}
+		deployTypeEnum, ok := usageeventsv1.UIDiscoverDeployServiceEvent_DeployType_value[d.ServiceDeploy.Type]
+		if !ok {
+			return nil, trace.BadParameter("invalid deployed type %s", d.ServiceDeploy.Type)
```
Comment (ravicious) [inline lib/usagereporter/web/userevent_discover.go:135]:
```suggestion
			return nil, trace.BadParameter("invalid deploy type %s", d.ServiceDeploy.Type)
```

**Review by ravicious — APPROVED** [review]

File: web/packages/teleport/src/services/userEvent/types.ts (hunk 4 above)
Comment (kimlisa) [reply web/packages/teleport/src/services/userEvent/types.ts:203]: we don't all cap acronyms atm in our frontend

**Pull request merged**